src/catchup/CatchupCompleteWork.cpp

```diff
@@ -76,8 +77,10 @@ CatchupCompleteWork::onSuccess()
                                  mCatchupTransactionsWork->getLastApplied())
                           << " for nextLedger=" << nextLedger();
     asio::error_code ec;
-    mEndHandler(ec, CatchupManager::CATCHUP_COMPLETE,
-                mCatchupTransactionsWork->getLastApplied());
+    mProgressHandler(ec, ProgressState::APPLIED_TRANSACTIONS,
+                     mCatchupTransactionsWork->getLastApplied());
+    mProgressHandler(ec, ProgressState::FINISHED,
```

Comment:
I could not really parse the comment in the commit for what this commit is trying to do; regardless this is an anti-pattern: callbacks of this type are expected to only be called once on completion, where as here it's called twice.

Reply:
This callback was already called twice in case of CATCHUP_RECENT, even that at this time it was called mEndHandler - once when it finished `inner CATCHUP_MINIMAL`, second time when it finished `inner CATCHUP_COMPLETE`. There was special logic inside `historyCaughtup` to take care of that:
>        // If we're in CATCHUP_RECENT mode, we actually only got part way
>        // through catchup -- we did the minimal prefix part -- and will
>        // get another callback as CATCHUP_COMPLETE when recent-replay is
>        // done. So for now just deposit LCL and prepare for replay.

I god rid of that with using a proper name - `mProgressHandler` and an explicit list of progress steps that will be reported to callback. This means that there always be at least 3 callbacks calls:

APPLIED_BUCKETS, FINISHED for CATCHUP_MINIMAL
APPLIED_TRANSACTIONS, FINISHED for CATCHUP_COMPLETE
APPLIED_BUCKETS, APPLIED_TRANSACTIONS, FINISHED for CATCHUP_RECENT

Reply:
sigh, at least it has a different type now. You should add comments to the definition of ProgressHandler that it the callback will be called in sequence with for example the contract in case of failure (it seems we call FINISHED with the ec set)